### PR TITLE
Fix wrong episode length after loading HER replay buffer

### PR DESCRIPTION
**Pull request by lbergmann1**
## Description

Fixes issue #1542 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: tests/test_her.py
Comment (qgallouedec) [inline tests/test_her.py:326]: I'd prefer a test dedicated to `truncate_last_trajectory`. In fact, the loading bug is simply a consequence of truncation not being performed correctly.
Something like: create a buffer, feed it with false interactions, truncate it, check that the warning is lifted (when it should be), feed it with a new interaction, and finally check that the attributes `observations`, `rewards`, ... are as expected.

File: docs/misc/changelog.rst
Comment (qgallouedec) [inline docs/misc/changelog.rst:44]: Following https://github.com/DLR-RM/stable-baselines3/pull/1543/files#r1219528110, it would be more accurate to say that `truncate_last_trajectory` has been fixed.

File: stable_baselines3/her/her_replay_buffer.py
Comment (lbergmann1) [reply stable_baselines3/her/her_replay_buffer.py:386]: If `self.dones[self.pos - 1] = False` before `truncate_last_trajectory`, the episode was not finished and `self.ep_length` is 0 for all entries belonging to this trajectory. Since `add()` is never called for this truncated trajectory, `self.ep_length[episode_indices]` is always 0 and cannot be sampled. So if I'm not mistaken, `self.dones[self.pos - 1] = True` has no effect.

Comment (qgallouedec) [reply stable_baselines3/her/her_replay_buffer.py:386]: You're right. I guess the question is therefore: do we wan't the truncated episode to be sampled? To be consistent with the previous version, I would say yes. So we've to keep this line, but also update `self.ep_length[self.pos - 1]`

Comment (lbergmann1) [reply stable_baselines3/her/her_replay_buffer.py:386]: What about infinite horizon tasks? If the episode is not finished before truncation, the done flag would become True too early.

Comment (qgallouedec) [reply stable_baselines3/her/her_replay_buffer.py:386]: Good point. The infinite horizon case is handled with the timeouts buffer. We should also update it, something like

```python
if self.handle_timeout_termination:
    self.timeouts[self.pos-1] = True
```

**Comment by qgallouedec** [comment]
Looks great!
Can you make the test independent of learning? Replace `model.learn` with a loop like this:

```python
observations = venv.reset()
for _ in range(4):
    actions = np.random.rand(n_envs, 4)
    next_observations, rewards, dones, infos = venv.step(actions)
    buffer.add(observations, next_observations, actions, rewards, dones, infos)
    observations = next_observations
```

**Comment by qgallouedec** [comment]
LGTM, I'll just wait for to CI to pass before merging. Thank you @lbergmann1!

**Review by qgallouedec — APPROVED** [review]

**Pull request merged**
